apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.RetryConformanceTests/RetryConformanceTest.cs

```diff
@@ -0,0 +1,370 @@
+
+    private readonly ConcurrentDictionary<string, MethodInvocation> _methodMappings;
+
+    internal string _hmacSecret;
+
+    internal string _notificationId;
+
+    public RetryConformanceTest()
+    {
+        _storageClient = InitializeClient();
+        _httpClient = new HttpClient
+        {
+            BaseAddress = new Uri(TestBenchUrl)
+        };
+
+        _methodMappings = new ConcurrentDictionary<string, MethodInvocation>();
+    }
+
+    [Theory]
+    [MemberData(nameof(RetryTestData))]
+    public async Task RetryTest(RetryTest test)
+    {
+        CreateMethodMapping(test);
+        bool expectSuccess = test.ExpectSuccess;
+
+        //Iterate throught the json file and run test cases
+        //UPDATE : Working for most test cases across except object related test cases and some exceptions to be fixed
+        foreach (InstructionList testCase in test.Cases)
+        {
+            JArray instruct = new JArray();
+            foreach (string instructions in testCase.Instructions)
+            {
+                instruct.Add(instructions);
+            }
+            foreach (Method method in test.Methods)
+            {
+                // _acl function dont exist in our library
+                if (!method.Name.Contains("_acl")) 
+                {
+                    await RunTestCase(instruct, method, expectSuccess, test.Description.Contains("when_precondition_is_present"));
+                }
+            }
+        }
+    }
+
+    private async Task RunTestCase(JArray instruction, Method method, bool expectSuccess, bool preConditionsPresent = false)
+    {
+        var resources = await CreateStorageResources(method);
+        var response = await CreateRetryTestResource(method, instruction);
+
+        bool success = true;
+        try
+        {
+            RunRetryTest(response, resources, preConditionsPresent);
+            var updatedResponse = await GetRetryTest(response.Id);
+            success = updatedResponse.Completed;
+        }
+        catch (System.Exception)
+        {
+            success = false;
+        }
+
+        Assert.Equal(expectSuccess, success);
+
+        // Delete the retry test and delete Storage Resource as well.
+        await DeleteRetryTest(response.Id);
+    }
+
+    private async Task<TestResponse> CreateRetryTestResource(Method method, JArray instruction)
+    {
+        var stringContent = GetBodyContent(method.Name, instruction);
+        HttpResponseMessage response = await _httpClient.PostAsync("retry_test", stringContent);
+        response.EnsureSuccessStatusCode();
+        var responseMessage = await response.Content.ReadAsStringAsync();
+        return JsonConvert.DeserializeObject<TestResponse>(responseMessage);
+    }
+
+    private void RunRetryTest(TestResponse response, StorageResources resources, bool preConditionsPresent)
+    {
+        AddRetryIdHeader(response.Id);
+
+        // Use method mapping to call the StorageTestBench API.
```

Comment:
Not a dotnet expert but I would expect this to go with the other code used to interact with the testbench -- I see you have TestResponse in a separate class/file, but it is possible to factor out all the code for interacting with the testbench so it's one abstraction basically?

Reply:
While sending the request, we use the C# Storage Client which internally takes care of the request. The response is serialized in a TestResponse class so that next set of actions can be taken based on the response. Could you please elaborate how do you want this as one abstraction. Thanks